### PR TITLE
Better user input sanitization: replace disableEveryone with disableMentions

### PR DESCRIPTION
**Pull request by y21**
**Please describe the changes this PR makes and why it should be merged:**
In an [earlier Pull Request](https://github.com/discordjs/discord.js/pull/3606), I tried fixing a bug where `disableEveryone` would not work if certain characters were in between the mention. At the time I was not aware that the RTL character (\u202e) can appear anywhere in the mention, which makes the Pull Request useless as it only works with `@\u202e?everyone`. Additionally, the Right to Left override character is not the only character that Discord strips.
A few months ago a [warning](https://discordapp.com/developers/docs/resources/channel#create-message) regarding message sanitization has been added to the API documentation:
> Discord may strip certain characters from message content, like invalid unicode characters or characters which cause unexpected message formatting. If you are passing user-generated strings into message content, consider sanitizing the data to prevent unexpected behavior. For example, to prevent unexpected mentions you could consider stripping @ characters or placing a zero-width space (U-200B) after them.

As it is unclear *what* characters specifically are stripped, it's rather hard to write a proper `disableEveryone` option (while still allowing other mentions such as user and role mentions). 
This Pull Request **removes** the ClientOption/MessageOption `disableEveryone` and adds `disableMentions`, which breaks **any** type of mention (except for channel mentions) instead of just `@everyone`/`@here` by adding a zero width space after every `@` character (as recommended by the API documentation).
Controlled mentions (such as user mentions at the beginning of message content in Message#reply) will not be affected by this change.

For clarification, I added tests that can be found in `test/disableMentions.js`. It shows that it is no longer possible to bypass the replacement using "certain characters":
__Using `disableMentions`:__
image
__Using `disableEveryone`:__
image

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/util/Util.js [hunk 1]
```diff
@@ -526,7 +526,7 @@ class Util {
    */
   static cleanContent(str, message) {
     return str
-      .replace(/@(everyone|here)/g, '@\u200b$1')
+      .replace(/@/g, '@\u200b')
       .replace(/<@!?[0-9]+>/g, input => {
```
Comment (NotSugden) [inline src/util/Util.js]: this regex will never be triggered as all `@`s will have a whitespace after them, perhaps add the whitespace later on?

File: src/structures/interfaces/TextBasedChannel.js [hunk 2]
```diff
@@ -57,8 +57,8 @@ class TextBasedChannel {
    * @property {string} [content=''] The content for the message
    * @property {MessageEmbed|Object} [embed] An embed for the message
    * (see [here](https://discordapp.com/developers/docs/resources/channel#embed-object) for more details)
-   * @property {boolean} [disableEveryone=this.client.options.disableEveryone] Whether or not @everyone and @here
-   * should be replaced with plain-text
+   * @property {boolean} [disableMentions=this.client.options.disableMentions] Whether or not mentions
+   * should be replaced with plain-text (this includes: role, user and everyone/here mentions)
```
Comment (NotSugden) [inline src/structures/interfaces/TextBasedChannel.js]: same as above

File: src/structures/Webhook.js [hunk 3]
```diff
@@ -85,8 +85,8 @@ class Webhook {
    * @property {string} [nonce=''] The nonce for the message
    * @property {Object[]} [embeds] An array of embeds for the message
    * (see [here](https://discordapp.com/developers/docs/resources/channel#embed-object) for more details)
-   * @property {boolean} [disableEveryone=this.client.options.disableEveryone] Whether or not @everyone and @here
-   * should be replaced with plain-text
+   * @property {boolean} [disableMentions=this.client.options.disableMentions] Whether or not mentions
+   * should be replaced with plain-text (this includes: role, user and everyone/here mentions)
```
Comment (NotSugden) [inline src/structures/Webhook.js]: shouldn't this state that it adds a whitespace after every @ character and not just mentions?

**Comment by y21** [comment]
Sorry, I forgot to push a commit that makes changes to `src/util/Util.js`. It adds a `removeMentions` method which breaks any type of mention (replaces @ with @\u200b). `Util.cleanContent` will use it if disableMentions option is present.

File: src/structures/Webhook.js (hunk 3 above)
Comment (y21) [reply src/structures/Webhook.js]: You are right, I should have been more clear here.

File: src/util/Util.js (hunk 1 above)
Comment (y21) [reply src/util/Util.js]: True. I changed it so it adds the zero width space *after* cleaning the message content.

File: src/util/Util.js [hunk 4]
```diff
@@ -518,6 +518,16 @@ class Util {
...
   }
 
+  /**
+   * Breaks user, role and everyone/here mentions by adding a zero width space after every @ character
+   * @param {string} str The string to sanitize
+   * @returns {string}
+   */
+  static removeMentions(str) {
+    return str
+      .replace(/@/g, '@\u200b');
```
Comment (amishshah) [inline src/util/Util.js]: This doesn't need to be on a newline

Comment (y21) [reply src/util/Util.js]: Made it one line. 👍

File: src/structures/Webhook.js [hunk 5]
```diff
@@ -85,8 +85,8 @@ class Webhook {
    * @property {string} [nonce=''] The nonce for the message
    * @property {Object[]} [embeds] An array of embeds for the message
    * (see [here](https://discordapp.com/developers/docs/resources/channel#embed-object) for more details)
-   * @property {boolean} [disableEveryone=this.client.options.disableEveryone] Whether or not @everyone and @here
-   * should be replaced with plain-text
+   * @property {boolean} [disableMentions=this.client.options.disableMentions]  Whether or not a zero width space
```
Comment (NotSugden) [inline src/structures/Webhook.js]:
```suggestion
   * @property {boolean} [disableMentions=this.client.options.disableMentions] Whether or not a zero width space
```
double space :eyes:

Comment (y21) [reply src/structures/Webhook.js]: oops

File: src/structures/APIMessage.js [hunk 6]
```diff
@@ -88,6 +88,13 @@ class APIMessage {
       content = Util.resolveString(this.options.content);
     }
 
+    const disableMentions = typeof this.options.disableMentions === 'undefined' ?
+      this.target.client.options.disableMentions :
+      this.options.disableMentions;
+    if (disableMentions) {
+      content = (content || '').replace(/@/g, '@\u200b');
```
Comment (amishshah) [inline src/structures/APIMessage.js]: Maybe you could use `Util.removeMentions` here?

Comment (y21) [reply src/structures/APIMessage.js]: Good idea.

File: src/util/Util.js [hunk 7]
```diff
@@ -550,7 +558,8 @@ class Util {
         if (message.channel.type === 'dm') return input;
         const role = message.guild.roles.cache.get(input.replace(/<|@|>|&/g, ''));
         return role ? `@${role.name}` : input;
-      });
+      })
+      .replace(/@/g, '@\u200b');
```
Comment (amishshah) [inline src/util/Util.js]: You could use `removeMentions` here too!

**Pull request merged**